File: Scripts/Data/DataPool/RelicPool.cs
using System.Collections.Generic;
using System.Linq;
using System;

namespace Roguelike.Data
{
    /// <summary>
    /// A data container that holds all possible RelicData templates for the game.
    /// Acts as a central database for relics
    /// </summary>
    public class RelicPool
    {
        /// <summary>
        /// The primary storage for all relics, indexed by their unique string ID for fast lookups.
        /// </summary>
        public Dictionary<string, RelicData> RelicsById { get; set; } = new Dictionary<string, RelicData>();
        
        /// <summary>
        /// A secondary index grouping relics by their star rating for quick access.
        /// </summary>
        public Dictionary<int, List<RelicData>> RelicsByStar { get; set; } = new Dictionary<int, List<RelicData>>();

        /// <summary>
        /// Defines the base shop cost for relics at each star rating.
        /// </summary>
        public Dictionary<int, int> BaseShopCosts { get; set; } = new Dictionary<int, int>();

        /// <summary>
        /// Initializes the RelicPool
        /// </summary>
        public void Initialize(IEnumerable<RelicData> allRelics)
        {
            RelicsById.Clear();
            RelicsByStar.Clear();

            foreach(var relic in allRelics)
            {
                RelicsById[relic.Id] = relic;
                
                if (!RelicsByStar.ContainsKey(relic.StarRating))
                {
                    RelicsByStar[relic.StarRating] = new List<RelicData>();
                }
                RelicsByStar[relic.StarRating].Add(relic);
            }

            BaseShopCosts.Clear();
            for (int i = 1; i <= 5; i++)
            {
                BaseShopCosts[i] = 60 * i;
            }
        }

        /// <summary>
        /// Retrieves a relic by its unique ID.
        /// </summary>
        public RelicData GetRelic(string id)
        {
            RelicsById.TryGetValue(id, out var relic);
            return relic;
        }

        /// <summary>
        /// Retrieves a random relic of the specified star rating.
        /// </summary>
        public RelicData GetRandomRelicOfStar(int star, System.Random rng)
        {
            if (RelicsByStar.TryGetValue(star, out var list) && list.Any())
            {
                return list[rng.Next(list.Count)];
            }
            return null;
        }
    }
}

File: python_roguelike/data/pools/relic_pool.py
import random
from typing import Dict, List, Optional
from ..relic_data import RelicData


class RelicPool:
    def __init__(self):
        self.relics_by_id: Dict[str, RelicData] = {}
        self.relics_by_star: Dict[int, List[RelicData]] = {}
        self.base_shop_costs: Dict[int, int] = {}

    def initialize(self, all_relics: List[RelicData]):
        self.relics_by_id.clear()
        self.relics_by_star.clear()

        for relic in all_relics:
            self.relics_by_id[relic.id] = relic
            if relic.star_rating not in self.relics_by_star:
                self.relics_by_star[relic.star_rating] = []
            self.relics_by_star[relic.star_rating].append(relic)

        self.base_shop_costs.clear()
        for i in range(1, 6):
            self.base_shop_costs[i] = 60 * i

    def get_relic(self, id: str) -> Optional[RelicData]:
        return self.relics_by_id.get(id)

    def get_random_relic_of_star(self, star: int, rng: random.Random) -> Optional[RelicData]:
        lst = self.relics_by_star.get(star)
        if lst:
            return rng.choice(lst)
        return None
